src/test/java/io/axoniq/axonserver/connector/query/QueryChannelIntegrationTest.java

```diff
@@ -343,6 +351,93 @@ public Registration registerSubscriptionQuery(SubscriptionQuery query, UpdateHan
+            queriesInProgress.add(reply);
+        }, new QueryDefinition("testQuery", String.class));
+
+        QueryChannel queryChannel2 = connection2.queryChannel();
+
+        assertWithin(1, TimeUnit.SECONDS, () -> connection1.isReady());
+
+        Thread.sleep(100); // this is because #5 (https://github.com/AxonIQ/axonserver-connector-java/issues/5) isn't implemented yet
+
+        List<ResultStream<QueryResponse>> actual = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            actual.add(queryChannel2.query(QueryRequest.newBuilder().setQuery("testQuery").build()));
+        }
+
+        assertWithin(1, TimeUnit.SECONDS, () -> assertEquals(10, queriesInProgress.size()));
+
+        ((ContextConnection) connection1).getManagedChannel().requestReconnect();
+        ((QueryChannelImpl) connection1.queryChannel()).reconnect();
+
+        assertWithin(1, TimeUnit.SECONDS, () -> assertTrue(connection1.isReady()));
+
+        while (!queriesInProgress.isEmpty()) {
+            doIfNotNull(queriesInProgress.poll(), r -> r.sendLast(QueryResponse.newBuilder().setPayload(SerializedObject.newBuilder().setType("java.lang.String").setData(ByteString.copyFromUtf8("Works"))).build()));
+        }
+
+        List<QueryResponse> actualMessages = actual.stream().map(r -> {
+            try {
+                return r.nextIfAvailable(1, TimeUnit.SECONDS);
+            } catch (InterruptedException e) {
+                return null;
+            }
+        }).collect(Collectors.toList());
+        actualMessages.forEach(r -> {
+            assertFalse(r.hasErrorMessage());
+        });
+
+    }
+
+    @Test
+    void testDisconnectFinishesQueriesInTransit() throws InterruptedException {
+        Queue<ReplyChannel<QueryResponse>> queriesInProgress = new ConcurrentLinkedQueue<>();
+        QueryChannel queryChannel = connection1.queryChannel();
+        queryChannel.registerQueryHandler((command, reply) -> {
+            CompletableFuture<QueryResponse> result = new CompletableFuture<>();
+            queriesInProgress.add(reply);
+        }, new QueryDefinition("testQuery", String.class));
+
+        QueryChannel queryChannel2 = connection2.queryChannel();
+
+        assertWithin(1, TimeUnit.SECONDS, () -> connection1.isReady());
+
+        Thread.sleep(100); // this is because #5 (https://github.com/AxonIQ/axonserver-connector-java/issues/5) isn't implemented yet
+
+        List<ResultStream<QueryResponse>> actual = new ArrayList<>();
+        for (int i = 0; i < 10; i++) {
+            actual.add(queryChannel2.query(QueryRequest.newBuilder().setQuery("testQuery").build()));
+        }
+
+        assertWithin(1, TimeUnit.SECONDS, () -> assertEquals(10, queriesInProgress.size()));
+
+        // a call to disconnect waits for disconnection to be completed. That isn't compatible with our test here.
+        CompletableFuture.runAsync(() -> connection1.disconnect());
```

Comment:
Can't we use the result of this operation to validate if the disconnect was successful at the end of this test?

Reply:
Added termination check at the end of the test.